src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpSamDAO.scala

```diff
@@ -5,61 +5,81 @@ import akka.http.scaladsl.Http
   private val samServiceURL = baseSamServiceURL
 
   val http = Http(system)
 
-  override def getStatus(): Future[SubsystemStatus] = {
+  override def getStatus(): Future[StatusCheckResponse] = {
     val uri = Uri(samServiceURL + "/status")
-    executeSamRequest(HttpRequest(GET, uri)).flatMap {
-      case HttpResponse(OK, _, entity, _) =>
-        Unmarshal(entity).to[SubsystemStatus]
-      case HttpResponse(status, _, entity, _) =>
-        val entityAsString = Unmarshal(entity).to[String]
-        Future.successful(SubsystemStatus(false, Some(List(s"Sam status check failed with code $status: $entityAsString"))))
+    executeSamRequest[StatusCheckResponse](HttpRequest(GET, uri), ignoreError = true) map { statusCheckResponse =>
+      // If it's an OK status, strip out any subsystem information from the Sam response.
+      // Otherwise, return the StatusCheckResponse as-is, with error information.
+      if (statusCheckResponse.ok) statusCheckResponse.copy(systems = Map.empty)
+      else statusCheckResponse
     }
   }
 
+  @deprecated(message = "Use getPetServiceAccountForProject(UserInfo, GoogleProject) instead", since = "")
   override def getPetServiceAccount(userInfo: UserInfo): Future[WorkbenchEmail] = {
     val uri = Uri(samServiceURL + "/api/user/petServiceAccount")
-    executeSamRequestAsUser(HttpRequest(GET, uri), userInfo).flatMap {
-      case HttpResponse(OK, _, entity, _) =>
-        Unmarshal(entity).to[WorkbenchEmail]
-      case HttpResponse(status, _, entity, _) =>
-        Unmarshal(entity).to[String].flatMap { entityAsString =>
-          Future.failed(CallToSamFailedException(uri, status, entityAsString))
-        }
-    }
+    executeSamRequestAsUser[WorkbenchEmail](HttpRequest(GET, uri), userInfo)
+  }
+
+  override def getPetServiceAccountForProject(userInfo: UserInfo, googleProject: GoogleProject): Future[WorkbenchEmail] = {
+    val uri = Uri(samServiceURL + s"/api/google/user/petServiceAccount/${googleProject.value}")
+    executeSamRequestAsUser[WorkbenchEmail](HttpRequest(GET, uri), userInfo)
   }
 
   private def authHeader(userInfo: UserInfo): HttpHeader = Authorization(userInfo.accessToken)
 
-  private def executeSamRequest(httpRequest: HttpRequest): Future[HttpResponse] = {
+  private def executeSamRequest[T](httpRequest: HttpRequest, ignoreError: Boolean = false)(implicit um: Unmarshaller[ResponseEntity, T]): Future[T] = {
     http.singleRequest(httpRequest) recover { case t: Throwable =>
-      throw CallToSamFailedException(httpRequest.uri, StatusCodes.InternalServerError, t.getMessage)
+      throw CallToSamFailedException(httpRequest.uri, StatusCodes.InternalServerError, Some(t.getMessage))
+    } flatMap { response =>
+      if (ignoreError || response.status.isSuccess) {
+        Unmarshal(response.entity).to[T]
+      } else {
+        // Try to unmarshal the response entity as an ErrorReport
+        Unmarshal(response.entity).to[ErrorReport] map { report =>
```

Comment:
Fixes another bugs. Before, when a Sam call failed and returned an `ErrorReport`, we deserialized the response entity as a `String` and returned the whole thing to the end user. Now we first try to deserialize the Sam error to an `ErrorReport`, and only return the message (not the huge stack trace) to the user. We still log the full `ErrorReport` for debugging.

Reply:
nice!

Reply:
When do we want to `ignoreError` ?

Reply:
In the case of status checks, a 500 status response contains the same entity as a 200 status (a `StatusCheckResponse`). So we set `ignoreError` to true in that case, and deserialize the requested type, instead of an `ErrorReport`.

I know, this is confusing.. will think about better names.